### PR TITLE
Feat/solana blockchain link

### PR DESCRIPTION
**Pull request by memoriesadrift**
## Description

Extracted solana blockchain-link changes into this PR as advised.

### CODE REVIEW

**Comment by mroz22** [comment]
> Any idea what's going on here?

I guess you did run `yarn`other than that I don't have a clue. 

Also please, could you reorganize commits to something meaning-ful please?

File: package.json [hunk 1]
```diff
@@ -37,8 +37,6 @@
         "_______ Global Scripts _______": "Shared scripts for running in all workspaces",
         "g:eslint": "cd $INIT_CWD && eslint --report-unused-disable-directives --cache --ignore-path ../../.gitignore",
         "g:jest": "cd $INIT_CWD && jest",
-        "g:rimraf": "cd $INIT_CWD && rimraf",
-        "g:tsc": "cd $INIT_CWD && tsc",
```
Comment (mroz22) [inline package.json]: this could be root of your problems in PR description

Comment (mroz22) [reply package.json]: in general, i would expect no changes to root package.json in your PR

File: packages/blockchain-link-utils/src/solana.ts [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+    TokenDetailByMint,
+    PublicKey,
+} from '@trezor/blockchain-link-types/lib/solana';
+import type { TokenInfo } from '@trezor/blockchain-link-types/lib';
+
+export type ApiTokenAccount = { account: AccountInfo<ParsedAccountData>; pubkey: PublicKey };
+
+export const TOKEN_PROGRAM_PUBLIC_KEY = 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA';
+export const ASSOCIATED_TOKEN_PROGRAM_PUBLIC_KEY = 'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL';
+export const SYSTEM_PROGRAM_PUBLIC_KEY = '11111111111111111111111111111111';
```
Comment (mroz22) [inline packages/blockchain-link-utils/src/solana.ts]: some comment or link to documentation on what are these constants would be helpful

File: packages/blockchain-link-utils/src/solana.ts [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+            ix.parsed.info?.mint === 'So11111111111111111111111111111111111111112',
+    );
+
+type TransactionEffect = {
+    address: string;
+    amount: BigNumber;
+};
+
+export function getNativeEffects(transaction: ParsedTransactionWithMeta): TransactionEffect[] {
+    // TODO(vl): after token PR is merged, get WSOL mint from there
```
Comment (mroz22) [inline packages/blockchain-link-utils/src/solana.ts]: could be that this Todo is outdated now?

File: packages/blockchain-link-utils/src/solana.ts [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+) => {
+    if (
+        effects.length === 1 &&
+        effects[0]?.address === accountAddress &&
+        effects[0]?.amount.abs().isEqualTo(new BigNumber(transaction.meta?.fee || 0))
+    ) {
+        return 'self';
+    }
+
+    const senders = effects.filter(({ amount }) => amount.isNegative());
```
Comment (mroz22) [inline packages/blockchain-link-utils/src/solana.ts]: nit pic: senders and receivers look like candidates to use `arrayPartition` utility

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+
+const BLOCK_SUBSCRIBE_INTERVAL_MS = 10000;
+const subscribeBlock = async ({ state, connect, post }: Context) => {
+    if (state.getSubscription('block')) return;
+    const api = await connect();
+
+    // the solana RPC api has subscribe method, see here: https://www.quicknode.com/docs/solana/rootSubscribe
+    // but solana block height is updated so often that it slows down the whole application and overloads the the api
+    // so we instead use setInterval to check for new blocks every 10 seconds
+    const interval = setInterval(async () => {
```
Comment (mroz22) [inline packages/blockchain-link/src/workers/solana/index.ts:286]: note: make sure that we don't register some unwanted intervals in case or repeated calls to this. maybe intervals should be cleared sometimes.

Comment (mroz22) [reply packages/blockchain-link/src/workers/solana/index.ts:286]: should be solved on line 280 here?

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+// https://github.com/viaprotocol/tokenlists
+// Aggregated token list with tokens listed on multiple exchanges
+const solanaTokenListUrl =
+    'https://cdn.jsdelivr.net/gh/viaprotocol/tokenlists/all_tokens/solana.json';
+
+export const getTokenMetadata = async (): Promise<TokenDetailByMint> => {
+    const tokenListResult: { address: string; name: string; symbol: string }[] = await (
+        await fetch(solanaTokenListUrl)
+    ).json();
```
Comment (mroz22) [inline packages/blockchain-link/src/workers/solana/tokenUtils.ts:29]: note: check how this is handled if fetch throws

File: packages/blockchain-link/tests/integration/solana.ts [hunk 7]
```diff
@@ -0,0 +1,210 @@
+import { AccountInfoParams } from '@trezor/blockchain-link-types';
```
Comment (mroz22) [inline packages/blockchain-link/tests/integration/solana.ts:1]: note for myself: check if and if yes how often we run these tests in CI.

**Comment by memoriesadrift** [comment]
> > Any idea what's going on here?
> 
> I guess you did run `yarn`other than that I don't have a clue.
> 
> Also please, could you reorganize commits to something meaning-ful please?

Alright, I'll get to it and force-push this branch afterwards.

File: package.json (hunk 1 above)
Comment (memoriesadrift) [reply package.json]: Reverted, everything works fine now.

File: packages/blockchain-link-utils/src/solana.ts (hunk 2 above)
Comment (memoriesadrift) [reply packages/blockchain-link-utils/src/solana.ts]: Added

File: packages/blockchain-link-utils/src/solana.ts (hunk 3 above)
Comment (memoriesadrift) [reply packages/blockchain-link-utils/src/solana.ts]: Yep, fixed.

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 5 above)
Comment (memoriesadrift) [reply packages/blockchain-link/src/workers/solana/index.ts:286]: cc: @PeterBenc

**Comment by memoriesadrift** [comment]
Commits reorganised, ready for review :)

File: packages/blockchain-link-types/package.json [hunk 8]
```diff
@@ -16,6 +16,7 @@
         "prepublish": "yarn tsx ../../scripts/prepublish.js"
     },
     "devDependencies": {
+        "@solana/web3.js": "1.78.4",
```
Comment (vytick) [inline packages/blockchain-link-types/package.json]: Hello, there seems to be multiple versions of `@solana/web3.js` -> **1.78.4** and **1.87.6**

File: packages/blockchain-link-utils/src/solana.ts [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+        preBalance: new BigNumber(preBalance ?? 0),
+        postBalance: new BigNumber(postBalance ?? 0),
+    };
+};
+
+const isWSolTransfer = (ixs: (ParsedInstruction | PartiallyDecodedInstruction)[]) =>
+    ixs.find(
+        ix =>
+            'parsed' in ix &&
+            ix.parsed.info?.mint === 'So11111111111111111111111111111111111111112',
```
Comment (vytick) [inline packages/blockchain-link-utils/src/solana.ts]: Maybe it will be good to use use `WSOL_MINT_LOWERCASE` or similar on all occurrences of wsol mint address? And have it defined on the same place where ata program and other addresses are defined?

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+                solanaUtils.transformTransaction(tx, payload.descriptor, slotToBlockHeightMapping),
+            )
+            .filter((tx): tx is Transaction => !!tx);
+    };
+
+    const tokenAccounts = await api.getParsedTokenAccountsByOwner(publicKey, {
+        programId: new PublicKey(TOKEN_PROGRAM_PUBLIC_KEY),
+    });
+
+    const allAccounts = [payload.descriptor, ...tokenAccounts.value.map(a => a.pubkey.toString())];
```
Comment (vytick) [inline packages/blockchain-link/src/workers/solana/index.ts:162]: I guess Stake Accounts are not part of this PR or out of scope, right?

File: packages/blockchain-link/src/workers/solana/index.ts [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+            break;
+        case 'accounts': {
+            unsubscribeAccounts(request, request.payload.accounts);
+            break;
+        }
+        default:
+            throw new CustomError('worker_unknown_request', `+${request.type}`);
+    }
+    return {
+        type: RESPONSES.SUBSCRIBE,
```
Comment (marekrjpolak) [inline packages/blockchain-link/src/workers/solana/index.ts]:
```suggestion
        type: RESPONSES.UNSUBSCRIBE,
```

**Review by vytick — COMMENTED** [review]
Hi guys, I was interested into Solana implementation so I looked at this :) . Just a few questions/findings which you can discard if not being relevant. No proper review :)

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 10 above)
Comment (memoriesadrift) [reply packages/blockchain-link/src/workers/solana/index.ts:162]: Yes, staking is out of scope for now.

**Review by marekrjpolak — CHANGES_REQUESTED** [review]

**Comment by marekrjpolak** [comment]
Also, is there a reason why solana-worker is defined in `workers.module.js` but not in `workers.web.js`?

**Comment by marekrjpolak** [comment]
Also, is there a reason why solana-worker is defined in `workers.module.js` but not in `workers.web.js`?

File: packages/blockchain-link/src/workers/solana/index.ts (hunk 5 above)
Comment (PeterBenc) [reply packages/blockchain-link/src/workers/solana/index.ts:286]: Yes @mroz22 should be solved by line 280, if there is a subscription already, we don't create another one. Also line 310 clears the interval. Do you consider this sufficient?

File: packages/blockchain-link/src/workers/solana/tokenUtils.ts (hunk 6 above)
Comment (memoriesadrift) [reply packages/blockchain-link/src/workers/solana/tokenUtils.ts:29]: As pointed out when originally discussing tokens, this is one of the possible solutions for handling token metadata, as such it wasn't necessarily meant to be a permanent solution. That's up to you to decide.
Quick recap on solutions:
1. Don't parse any token metadata, just use token mint address
2. Maintain local token whitelist
3. Fetch external token list (currently in use here)

**Comment by memoriesadrift** [comment]
> Also, is there a reason why solana-worker is defined in `workers.module.js` but not in `workers.web.js`?

Added, was an oversight.

**Review by mroz22 — APPROVED** [review]
ok, #9962 seems to run CI well

**Comment by mroz22** [comment]
https://github.com/trezor/trezor-suite/pull/9962 merged
